astropy/io/ascii/tests/test_read.py

```diff
@@ -1183,3 +1183,26 @@ def test_no_units_for_char_columns():
     ascii.write(t1, out, format="ipac")
     t2 = ascii.read(out.getvalue(), format="ipac", guess=False)
     assert t2["B"].unit is None
+
+
+def test_initial_column_fill_values():
+    """Regression test for #5336, #5338."""
+
+    class TestHeader(ascii.BasicHeader):
+        def _set_cols_from_names(self):
+            self.cols = [ascii.Column(name=x) for x in self.names]
+            # Set some initial fill values
+            for i in self.cols:
+                i.fill_values = {'--': 0}
```

Comment:
This should be `{'--', '0'}`.  I.e. the second element should be a string because it is substituted in the part of processing where the column exists as a list of strings.  The fact that it works (assuming tests pass) is slightly accidental.  See the definition of `missing_spec` in http://astropy.readthedocs.io/en/latest/io/ascii/read.html#bad-or-missing-values for details.